libraries/Adapters/Microsoft.Bot.Builder.Adapters.WeChat/MessageCryptography.cs

```diff
@@ -0,0 +1,168 @@
+            _appId = secretInfo.AppId;
+            _encodingAesKey = secretInfo.EncodingAESKey;
+            _msgSignature = secretInfo.Msg_Signature;
+            _timestamp = secretInfo.Timestamp;
+            _nonce = secretInfo.Nonce;
+        }
+
+        /// <summary>
+        /// Verify the authenticity of the message and get the decrypted plaintext.
+        /// </summary>
+        /// <param name="postData">cipher message.</param>
+        /// <returns>Decrypted message string.</returns>
+        public string DecryptMessage(string postData)
+        {
+            if (_encodingAesKey.Length != 43)
+            {
+                throw new ArgumentException("Invalid EncodingAESKey");
+            }
+
+            // This should fix the XXE loophole
+            var doc = new XmlDocument
+            {
+                XmlResolver = null,
+            };
+            XmlNode root;
+            string encryptMessage;
+            try
+            {
+                doc.LoadXml(postData);
+                root = doc.FirstChild;
+                encryptMessage = root["Encrypt"].InnerText;
+            }
+            catch (Exception e)
+            {
+                throw e;
+            }
+
+            if (!VerificationHelper.VerifySignature(_msgSignature, _token, _timestamp, _nonce, encryptMessage))
+            {
+                throw new UnauthorizedAccessException("Signature validation failed.");
+            }
+
+            return AesDecrypt(encryptMessage, _encodingAesKey, _appId);
+        }
+
+        /// <summary>
+        /// Decrypt the message.
+        /// </summary>
+        /// <param name="encryptString">Encrypted string.</param>
+        /// <param name="encodingAesKey">Encoding AES key for descrypt message.</param>
+        /// <param name="appId">The WeChat app id.</param>
+        /// <returns>Decrypted string.</returns>
+        public string AesDecrypt(string encryptString, string encodingAesKey, string appId)
+        {
+            try
+            {
+                var key = Convert.FromBase64String(encodingAesKey + "=");
+                var iv = new byte[16];
+                Array.Copy(key, iv, 16);
+                var btmpMsg = AesDecrypt(encryptString, iv, key);
+
+                var len = BitConverter.ToInt32(btmpMsg, 16);
+                len = IPAddress.NetworkToHostOrder(len);
+
+                var messageBytes = new byte[len];
+                var appIdBytes = new byte[btmpMsg.Length - 20 - len];
+                Array.Copy(btmpMsg, 20, messageBytes, 0, len);
+                Array.Copy(btmpMsg, 20 + len, appIdBytes, 0, btmpMsg.Length - 20 - len);
+                var oriMsg = Encoding.UTF8.GetString(messageBytes);
+                if (appId != Encoding.UTF8.GetString(appIdBytes))
+                {
+                    throw new ArgumentException("Failed to validate appId.", nameof(appId));
+                }
+
+                return oriMsg;
+            }
+            catch (FormatException e)
```

Comment:
Consider removing this catch block, it is not adding much value, just let the original exception flow up if it happens.